update a service with --restart-delay causes an error

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

For more information about reporting issues, see
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#reporting-other-issues

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Output of `docker version`:**

```
ubuntu@manager:~$ docker service update  --restart-delay 2s dm
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x74b67d]

goroutine 1 [running]:
panic(0xbeec40, 0xc82000e0e0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/docker/docker/api/client/service.updateService.func6(0xd18e60, 0xd, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/api/client/service/update.go:115 +0xad
github.com/docker/docker/api/client/service.updateService(0xc820076ab0, 0xc8200ca3c8, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/api/client/service/update.go:174 +0x6ef
github.com/docker/docker/api/client/service.runUpdate(0xc820288b80, 0xc820076ab0, 0x7fff90a1c8f3, 0x2, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/api/client/service/update.go:50 +0x1f9
github.com/docker/docker/api/client/service.newUpdateCommand.func1(0xc820304000, 0xc820320e70, 0x1, 0x3, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/api/client/service/update.go:28 +0x6e
github.com/spf13/cobra.(*Command).execute(0xc820304000, 0xc820320de0, 0x3, 0x3, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/spf13/cobra/command.go:593 +0x6fe
github.com/spf13/cobra.(*Command).ExecuteC(0xc8200798c0, 0xc820304000, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/spf13/cobra/command.go:683 +0x55c
github.com/spf13/cobra.(*Command).Execute(0xc8200798c0, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/spf13/cobra/command.go:642 +0x2d
github.com/docker/docker/cli/cobraadaptor.CobraAdaptor.run(0xc8200798c0, 0xc820288b80, 0x7fff90a1c8d1, 0x7, 0xc82000a140, 0x4, 0x4, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/cobraadaptor/adaptor.go:117 +0x25f
github.com/docker/docker/cli/cobraadaptor.CobraAdaptor.Command.func1(0xc82000a140, 0x4, 0x4, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/cobraadaptor/adaptor.go:125 +0x8c
github.com/docker/docker/cli.(*Cli).Run(0xc820320870, 0xc82000a130, 0x5, 0x5, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/cli/cli.go:81 +0x34b
main.main()
	/usr/src/docker/cmd/docker/docker.go:72 +0x599
```


**Output of `docker info`:**

```
buntu@manager:~$ docker info
Containers: 11
 Running: 1
 Paused: 0
 Stopped: 10
Images: 4
Server Version: 1.12.0-rc4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 48
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: active
 NodeID: 9i69l7phlv8lzvh4beung1l8q
 IsManager: Yes
 Managers: 1
 Nodes: 3
 CACertHash: sha256:f1e4f751c598a9b6aa8999cf1f5797539c68d98cab5985c449900e6b701c60e7
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-92-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 992.5 MiB
Name: ip-10-0-0-59
ID: USR7:ZEWH:3VLG:HI23:AUIQ:EGG4:ARLG:NNAR:4RO2:7PFY:LIZJ:UPRX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: ironsrcci
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS

--- a/api/client/service/update.go
+++ b/api/client/service/update.go
@@ -98,7 +98,6 @@ func runUpdate(dockerCli *client.DockerCli, flags *pflag.FlagSet, serviceID stri
 }
 
 func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
-
 	updateString := func(flag string, field *string) {
 		if flags.Changed(flag) {
 			*field, _ = flags.GetString(flag)
@@ -117,9 +116,10 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
 		}
 	}
 
-	updateDurationOpt := func(flag string, field *time.Duration) {
+	updateDurationOpt := func(flag string, field **time.Duration) {
 		if flags.Changed(flag) {
-			*field = *flags.Lookup(flag).Value.(*DurationOpt).Value()
+			val := *flags.Lookup(flag).Value.(*DurationOpt).Value()
+			*field = &val
 		}
 	}
 
@@ -129,9 +129,10 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
 		}
 	}
 
-	updateUint64Opt := func(flag string, field *uint64) {
+	updateUint64Opt := func(flag string, field **uint64) {
 		if flags.Changed(flag) {
-			*field = *flags.Lookup(flag).Value.(*Uint64Opt).Value()
+			val := *flags.Lookup(flag).Value.(*Uint64Opt).Value()
+			*field = &val
 		}
 	}
 
@@ -159,7 +160,6 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
 		taskResources().Limits = &swarm.Resources{}
 		updateInt64Value(flagLimitCPU, &task.Resources.Limits.NanoCPUs)
 		updateInt64Value(flagLimitMemory, &task.Resources.Limits.MemoryBytes)
-
 	}
 	if flags.Changed(flagReserveCPU) || flags.Changed(flagReserveMemory) {
 		taskResources().Reservations = &swarm.Resources{}
@@ -167,7 +167,7 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
 		updateInt64Value(flagReserveMemory, &task.Resources.Reservations.MemoryBytes)
 	}
 
-	updateDurationOpt(flagStopGracePeriod, cspec.StopGracePeriod)
+	updateDurationOpt(flagStopGracePeriod, &cspec.StopGracePeriod)
 
 	if anyChanged(flags, flagRestartCondition, flagRestartDelay, flagRestartMaxAttempts, flagRestartWindow) {
 		if task.RestartPolicy == nil {
@@ -178,9 +178,9 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
 			value, _ := flags.GetString(flagRestartCondition)
 			task.RestartPolicy.Condition = swarm.RestartPolicyCondition(value)
 		}
-		updateDurationOpt(flagRestartDelay, task.RestartPolicy.Delay)
-		updateUint64Opt(flagRestartMaxAttempts, task.RestartPolicy.MaxAttempts)
-		updateDurationOpt((flagRestartWindow), task.RestartPolicy.Window)
+		updateDurationOpt(flagRestartDelay, &task.RestartPolicy.Delay)
+		updateUint64Opt(flagRestartMaxAttempts, &task.RestartPolicy.MaxAttempts)
+		updateDurationOpt(flagRestartWindow, &task.RestartPolicy.Window)
 	}
 
 	if anyChanged(flags, flagConstraintAdd, flagConstraintRemove) {
@@ -206,6 +206,9 @@ func updateService(flags *pflag.FlagSet, spec *swarm.ServiceSpec) error {
 	updateNetworks(flags, &spec.Networks)
 	if flags.Changed(flagEndpointMode) {
 		value, _ := flags.GetString(flagEndpointMode)
+		if spec.EndpointSpec == nil {
+			spec.EndpointSpec = &swarm.EndpointSpec{}
+		}
 		spec.EndpointSpec.Mode = swarm.ResolutionMode(value)
 	}
 
@@ -409,7 +412,7 @@ func updateReplicas(flags *pflag.FlagSet, serviceMode *swarm.ServiceMode) error 
 		return nil
 	}
 
-	if serviceMode.Replicated == nil {
+	if serviceMode == nil || serviceMode.Replicated == nil {
 		return fmt.Errorf("replicas can only be used with replicated mode")
 	}
 	serviceMode.Replicated.Replicas = flags.Lookup(flagReplicas).Value.(*Uint64Opt).Value()

